src/System.Private.CoreLib/src/System/GC.cs

```diff
@@ -508,5 +513,30 @@ public static void EndNoGCRegion()
         {
             EndNoGCRegionWorker();
         }
+
+        // Skips zero-initialization of the array if possible. If T contains object references, 
+        // the array is always zero-initialized.
+        internal static T[] AllocateUninitializedArray<T>(int length)
+        {
+            if (length < 0)
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.lengths, 0, ExceptionResource.ArgumentOutOfRange_NeedNonNegNum);
+
+#if DEBUG
+            // in DEBUG arrays of any length can be created uninitialized
+#else
+            // otherwise small arrays are allocated using `new[]` as that is generally faster.
+            //
+            // The threshold was derived from various simulations. 
+            // As it turned out the threshold depends on overal pattern of all allocations and is typically in 200-300 byte range.
+            // The gradient around the number is shallow (there is no perf cliff) and the exact value of the threshold does not matter a lot.
```

Comment:
What was the test that you have used to decide the threshold? `256 - 3 * IntPtr.Size` is much smaller than what I would expect.

Reply:
GCPerfSim + GC perf infrastructure.  There are more details to this. 

---
In reply to: [279238932](https://github.com/dotnet/coreclr/pull/24096#discussion_r279238932) [](ancestors = 279238932)